File: LeetCode/Python/Hard/446-Arithmetic-Slices-II-Subsequence/solution.py
'''
Solution 1: pure recursion. DO NOT submit this solution as it will get Time Limit Exceeded.
'''
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        result=0

        n=len(nums)

        hashmap=dict()

        for (index, number) in enumerate(nums):
            if number not in hashmap.keys():
                hashmap[number]=list()
            
            hashmap[number].append(index)

        for i in range(0, n):
            for j in range(i + 1, n):
                difference=nums[j] - nums[i]

                target=nums[j] + difference

                result+=self.helper(nums, hashmap, n, difference, target, j)

        return result

    def helper(self, nums: List[int], hashmap: Dict[int, List[int]], n: int, difference: int, target: int, index: int) -> int:
        if index==n:
            return 0

        if target not in hashmap.keys():
            return 0

        subproblem_solution=0

        for i in hashmap[target]:
            if i > index:
                next_target=nums[i] + difference

                subproblem_solution+=1 + self.helper(nums, hashmap, n, difference, next_target, i)

        return subproblem_solution

'''
Solution 2: recursion + memoization. This solution is accepted.
'''
class Solution:
    def __init__(self):
        self.memo=dict()

    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        result=0

        n=len(nums)

        hashmap=dict()

        for (index, number) in enumerate(nums):
            if number not in hashmap.keys():
                hashmap[number]=list()
            
            hashmap[number].append(index)

        for i in range(0, n):
            for j in range(i + 1, n):
                difference=nums[j] - nums[i]

                target=nums[j] + difference

                result+=self.helper(nums, hashmap, n, difference, target, j)

        return result

    def helper(self, nums: List[int], hashmap: Dict[int, List[int]], n: int, difference: int, target: int, index: int) -> int:
        if index==n:
            return 0

        if target not in hashmap.keys():
            return 0

        key=(target, index)

        if key in self.memo.keys():
            return self.memo[key]

        subproblem_solution=0

        for i in hashmap[target]:
            if i > index:
                next_target=nums[i] + difference

                subproblem_solution+=1 + self.helper(nums, hashmap, n, difference, next_target, i)

        self.memo[key]=subproblem_solution

        return subproblem_solution

File: LeetCode/Rust/Hard/446-Arithmetic-Slices-II-Subsequence/solution.rs
use std::collections::HashMap;

impl Solution
{
    pub fn number_of_arithmetic_slices(nums: Vec<i32>) -> i32
    {
        let mut result: i32=0;

        let n: usize=nums.len();

        let mut hashmap: HashMap<i64, Vec<usize>>=HashMap::new();

        for (index, number) in nums.iter().enumerate()
        {
            hashmap.entry(*number as i64).or_insert(vec![]).push(index);
        }

        let mut memo: HashMap<(i64, usize), i32>=HashMap::new();

        for i in 0..n
        {
            for j in i + 1..n
            {
                let difference: i64=nums[j] as i64 - nums[i] as i64;

                let target: i64=nums[j] as i64 + difference;

                result+=Self::helper(&nums, &mut memo, &hashmap, n, difference, target, j);
            }
        }     

        result
    }

    fn helper(nums: &Vec<i32>, memo: &mut HashMap<(i64, usize), i32>, hashmap: &HashMap<i64, Vec<usize>>, n: usize, difference: i64, target: i64, index: usize) -> i32
    {
        if index==n
        {
            return 0;
        }

        if !hashmap.contains_key(&target)
        {
            return 0;
        }

        let key: (i64, usize)=(target, index);

        if memo.contains_key(&key)
        {
            return *memo.get(&key).unwrap();
        }

        let mut subproblem_solution: i32=0;

        for i in hashmap.get(&target).unwrap().iter()
        {
            if *i > index
            {
                let next_target: i64=nums[*i] as i64 + difference;
                
                subproblem_solution+=1 + Self::helper(nums, memo, hashmap, n, difference, next_target, *i);
            }
        }

        memo.insert(key, subproblem_solution);

        subproblem_solution
    }
}